rst_files/wald_friedman.rst

```diff
@@ -401,467 +345,201 @@ that appear on the right side of the Bellman equation as separate functions
-        m, pgrid = wf.m, wf.pgrid
-
-        J_out = similar(J)
-        J_interp = LinearInterpolation(pgrid, J)
-
-        for (p_ind, p) in enumerate(pgrid)
-            # Payoff of choosing model 0
-            p_c_0 = payoff_choose_f0(wf, p)
-            p_c_1 = payoff_choose_f1(wf, p)
-            p_con = payoff_continue(wf, p, J_interp)
-
-            J_out[p_ind] = min(p_c_0, p_c_1, p_con)
+    function decision_rule(d0, d1, L0, L1, c)
+        function cost(p, d0, d1, L0, L1, c)
+            c ≥ zero(c) || throw(ArgumentError("Cost must be non-negative"))
+            x0 = accept_x0(p, L0)
+            x1 = accept_x1(p, L1)
+            draw = draw_again(p, d0, d1, L0, L1, c, min(x0, x1))
+            output = min(abs(draw - x0), abs(draw - x1), abs(x1 - x0))
+            return output
         end
-
-        return J_out
-    end
-
-    function find_cutoff_rule(wf, J)
-        m, pgrid = wf.m, wf.pgrid
-
-        # Evaluate cost at all points on grid for choosing a model
-        p_c_0 = payoff_choose_f0.(Ref(wf), pgrid)
-        p_c_1 = payoff_choose_f1.(Ref(wf), pgrid)
-
-        # The cutoff points can be found by differencing these costs with
-        # the Bellman equation (J is always less than or equal to p_c_i)
-        lb = pgrid[searchsortedlast(p_c_1 - J, 1e-10)]
-        ub = pgrid[searchsortedlast(J - p_c_0, -1e-10)]
-
-        return lb, ub
-    end
-
-    function solve_model!(wf; tol = 1e-7)
-        bell_op(x) = bellman_operator(wf, x)
-        J =  compute_fixed_point(bell_op, zeros(wf.m), err_tol=tol, print_skip=5)
-
-        wf.sol.J = J
-        wf.sol.lb, wf.sol.ub = find_cutoff_rule(wf, J)
-        return J
-    end
-
-    function simulate(wf, f; p0 = 0.5)
-        # Check whether vf is computed
-        if sum(abs, wf.sol.J) < 1e-8
-            solve_model!(wf)
+        roots = find_zeros(p -> cost(p, d0, d1, L0, L1, c), 0, 1 - eps())
+        left = first.(choice.(roots .- eps(), d0, d1, L0, L1, c))
+        right = first.(choice.(roots .+ eps(), d0, d1, L0, L1, c))
+        β, α = (0.0, 1.0)
+        for (left, right, root) in zip(left, right, roots)
+            if left == 2 && right ≠ 2
+                β = root
+            end
         end
-
-        # Unpack useful info
-        lb, ub = wf.sol.lb, wf.sol.ub
-        drv = DiscreteRV(f)
-
-        # Initialize a couple useful variables
-        decision = 0
-        p = p0
-        t = 0
-
-        while true
-            # Maybe should specify which distribution is correct one so that
-            # the draws come from the "right" distribution
-            k = rand(drv)
-            t = t + 1
-            p = bayes_update_k(wf, p, k)
-            if p < lb
-                decision = 1
-                break
-            elseif p > ub
-                decision = 0
-                break
+        for (left, right, root) in zip(reverse(left), reverse(right), reverse(roots))
+            if left ≠ 1 && right == 1
```

Comment:
Use ascii for control flow

Reply:
Replace `&&` with `&`? Or `≠` with `!=`? `≠` is closer to the math.

Reply:
Those the style also mean I should change `≥`/`≤` for control flow? Those are the other non-ASCII used in control flow in this lectures I believe.

Reply:
Yeah, I think that is right.  Though I really do love the unicode for ≠ if we are writing up a line of math where it is used for set inclusion, etc.

Reply:
For set inclusion it would be `∈` or `∉`. I don't think `≠` is used outside control flow or tests... Very few exceptions is any. Probably best to discuss it later on the style guide, but I would cast my vote towards keeping control flow operations close to the math as well.

Reply:
Of course, you are right.  Then I don't think there is any reason to use ≠  except when doing DSL-y type stuff

Control flow is part of an algorithm, and not a direct implementation of a mathematical formula in the lecture notes.  Making the distinction between the direct implementation of the formulas from the math and the control flow of algorithms is a key goal.